src/Sparrow/Json/JsonContextPoolBase.cs

```diff
@@ -152,57 +182,95 @@ public void Dispose()
+                var currentGlobalStack = _currentGlobalStack;
+
+                var perCoreCleanupNeeded = currentGlobalStack.IsEmpty || currentTime - _lastPerCoreCleanup >= _perCoreCleanupInterval;
+                if (perCoreCleanupNeeded)
                 {
-                    for (var index = 0; index < current.Length; index++)
+                    _lastPerCoreCleanup = currentTime;
+
+                    foreach (var current in _perCoreContexts)
                     {
-                        var context = current[index];
-                        if(context == null)
-                            continue;
+                        for (var index = 0; index < current.Length; index++)
+                        {
+                            var context = current[index];
+                            if (context == null)
+                                continue;
+
+                            var timeInPool = currentTime - context.InPoolSince;
+                            if (timeInPool <= idleTime)
+                                continue;
+
+                            if (context.InUse.Raise() == false)
+                                continue;
+
+                            Interlocked.CompareExchange(ref current[index], null, context);
+                            context.Dispose();
+                        }
+                    }
 
-                        if (currentTime - context.InPoolSince > idleTime)
-                            continue;
+                    return;
+                }
 
-                        if (!context.InUse.Raise())
+                using (var globalStackEnumerator = currentGlobalStack.GetEnumerator())
+                {
+                    while (globalStackEnumerator.MoveNext())
+                    {
+                        var context = globalStackEnumerator.Current;
+
+                        var timeInPool = currentTime - context.InPoolSince;
+                        if (timeInPool <= idleTime)
                             continue;
 
-                        Interlocked.CompareExchange(ref current[index], null, context);
-                        context.Dispose();
+                        if (context.InUse.Raise())
+                            context.Dispose();
                     }
                 }
 
-                return;
-            }
+                var globalStackRebuildNeeded = currentTime - _lastGlobalStackRebuild >= _globalStackRebuildInterval;
 
-            while (_globalQueue.TryPeek(out var context))
-            {
-                if (currentTime - context.InPoolSince < idleTime)
+                if (globalStackRebuildNeeded)
                 {
-                    break;
+                    _lastGlobalStackRebuild = currentTime;
+
+                    var previousGlobalStack = _previousGlobalStack;
+
+                    _currentGlobalStack = new CountingConcurrentStack<T>();
+                    _previousGlobalStack = currentGlobalStack;
```

Comment:
Isn't this a problem when we are running concurrently?
Other threads may add to their view of `_currentGlobalStack` after you called `ClearStack`

Reply:
And why not simply clearing the *current* stack and be done with it?

Reply:
The idea here is that I still want to retain (maybe unnecessary?) contexts from the global stack that are not outdated. I'm doing that by creating new stack, but when allocating I'm taking values first from the 'old stack', then releasing them to the 'current stack'. I'm doing that, because I'm not removing anything from the stack during the cleanup (regular one - not the one that happens every 15min), only disposing contexts there.

The other option would be just to release everything as you wrote. It is easier, but not sure if we want that?

Reply:
The easiest thing to do would be to keep the queue structure, and scan the whole queue.
If there are old stuff there, just mark it in use and dispose it.
The null pull from the queue will clear it up, no?